DesiredBalance: expose it via _internal/desired_balance

arteam opened the pull request:
Add an internal endpoint for exposed the desired balance and computation stats at a master node as `GET _internal/desired_balance` and returns

```json
{
  "stats": {
    "computation_active": false,
    "computation_submitted": 5,
    "computation_executed": 5,
    "computation_converged": 5,
    "computation_iterations": 4,
    "computation_converged_index": 4,
    "computation_time_in_millis": 0,
    "reconciliation_time_in_millis": 0
  },
  "routing_table": {
    "test": {
      "0": {
        "current": [
          {
            "state": "STARTED",
            "primary": true,
            "node": "UPYt8VwWTt-IADAEbqpLxA",
            "node_is_desired": true,
            "relocating_node": null,
            "relocating_node_is_desired": false,
            "shard_id": 0,
            "index": "test"
          }
        ],
        "desired": {
          "node_ids": [
            "UPYt8VwWTt-IADAEbqpLxA"
          ],
          "total": 1,
          "unassigned": 0,
          "ignored": 0
        }
      },
      "1": {
        "current": [
          {
            "state": "STARTED",
            "primary": true,
            "node": "2x1VTuSOQdeguXPdN73yRw",
            "node_is_desired": true,
            "relocating_node": null,
            "relocating_node_is_desired": false,
            "shard_id": 1,
            "index": "test"
          }
        ],
        "desired": {
          "node_ids": [
            "2x1VTuSOQdeguXPdN73yRw"
          ],
          "total": 1,
          "unassigned": 0,
          "ignored": 0
        }
      }
    }
  }
}
```


Fixes #90583

arteam committed:
DesiredBalance: expose it via _internal/desired_balance

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,111 @@
+        Map<ShardId, ShardAssignment> assignments = latestDesiredBalance.assignments();
+        for (Map.Entry<ShardId, ShardAssignment> entry : assignments.entrySet()) {
+            ShardAssignment shardAssignment = entry.getValue();
+            // TODO Do we have to return only the primary shard or replicas as well?
```
ShardAssignment contains nodes for the primary and all replicas.

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,111 @@
+        DesiredNodes desiredNodes = DesiredNodesMetadata.fromClusterState(state).getLatestDesiredNodes();
+        Map<String, Map<Integer, DesiredBalanceResponse.DesiredShards>> routingTable = new HashMap<>();
+        Map<ShardId, ShardAssignment> assignments = latestDesiredBalance.assignments();
+        for (Map.Entry<ShardId, ShardAssignment> entry : assignments.entrySet()) {
```
I wonder if we should iterate over the routing nodes instead?

We need to keep in mind unassigned shards that might have no balance (in case the index was just created and computation is not yet finished)

arteam replied:
Yeah, that was one of the earlier ideas. I am going to change it to iterating over the routing nodes.

arteam committed:
Iterate over the routing table for desired shards

arteam committed:
Add REST tests

arteam committed:
Remove System.out.printlin

arteam committed:
Fix the docs example for getting desired nodes

arteam committed:
Comment out computation_active

Can't predict whether a node is desired

arteam committed:
Fix docs

Remove unused imports

Remove TBD from docs

arteam committed:
Use correct test conventions

Fix naming conventions for DesiredBalanceResponseTests

arteam committed:
Comment out match for node ids

arteam committed:
Add additional routing table checks

arteam committed:
Comment out node ids assertions

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

idegtiarenko commented on docs/reference/cluster/get-desired-balance.asciidoc:
```diff
@@ -0,0 +1,77 @@
+					"node": "UPYt8VwWTt-IADAEbqpLxA",
+					"node_is_desired": false,
+					"relocating_node": null,
+					"relocating_node_is_desired": false,
```
I wonder if we want to show above 2 fields in case we do not reallocate? Both true/false in `relocating_node_is_desired`  are some what confusing if we are not relocating. If we always want to show all fields then may be we should consider null value if we are not relocating?

arteam replied:
I think it's generally good idea to keep all the fields in the JSON representations of objects which makes it easier for clients to consume them since they don't need to do additional check on the existence of the fields. 

To your second point, I kind of prefer to avoid three-valued logic for boolean fields, unless there's substantial semantic difference between `null` and `false`. I feel in this case if the relocating node is null, so it can't be desired, thus `relocating_node_is_desired` being `false` makes sense.

idegtiarenko commented on docs/reference/cluster/get-desired-balance.asciidoc:
```diff
@@ -0,0 +1,77 @@
+					"state": "STARTED",
+					"primary": true,
+					"node": "UPYt8VwWTt-IADAEbqpLxA",
+					"node_is_desired": false,
```
Not sure if it is just a sample or implementation issue as well, but this should be true as this value is contained in `node_ids` in `desired` object below. Anyways I would suggest to change to true if this is presented as a doc

arteam replied:
Good call!

idegtiarenko commented on rest-api-spec/src/main/resources/rest-api-spec/api/_internal.get_desired_balance.json:
```diff
@@ -0,0 +1,23 @@
+  "_internal.get_desired_balance":{
+    "documentation":{
+      "url":"https://www.elastic.co/guide/en/elasticsearch/reference/current/get-desired-balance.html",
+      "description": "Gets the latest desired balance and basic metrics. Designed for indirect use by ECE/ESS and ECK. Direct use is not supported."
```
I believe this is common for an internal api to mention ECE/ESS. Though I am not sure we should do it here as well.

arteam replied:
I used this just a stock disclaimer for internal APIs. Would some different wording better here?

henningandersen replied:
We can keep it experimental.

Also, we can instead say something like

"This API is a diagnostics API and the output should not be relied upon for building applications."

No need to mention the cloud/ece/eck, desired balance is for all.

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/DesiredBalanceResponse.java:
```diff
@@ -0,0 +1,207 @@
+        return routingTable;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        return o instanceof DesiredBalanceResponse that
+            && Objects.equals(stats, that.stats)
+            && Objects.equals(routingTable, that.routingTable);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(stats, routingTable);
+    }
+
+    @Override
+    public String toString() {
+        return "DesiredBalanceResponse{stats=" + stats + ", routingTable=" + routingTable + "}";
+    }
```
NIT: DesiredBalanceResponse could be a record too

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,111 @@
+        ClusterState state,
+        ActionListener<DesiredBalanceResponse> listener
+    ) throws Exception {
+        DesiredBalance latestDesiredBalance = desiredBalanceShardsAllocator.getDesiredBalance();
```
I think we also need to handle the case if the cluster is configured to use `cluster.routing.allocation.type=balanced` and show a corresponding message.

idegtiarenko commented on server/src/test/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceActionTests.java:
```diff
@@ -0,0 +1,65 @@
+        assertEquals(shard.currentNodeId(), desiredShards.current().node());
+        assertEquals(shard.relocatingNodeId(), desiredShards.current().relocatingNode());
+        assertFalse(desiredShards.current().relocatingNodeIsDesired());
+        // Desired balance isn't stable, unable to reliably make mode ids assertions
```
I think it might be reasonable to use a 1 or 2 shard index pinned to a specific nodes with `routing.allocation.include=node-id` for the purpose of making a test more predictable.

`routing.allocation.include=fake` could be used to check a result for a shard that could not be assigned.

It should be possible to fake a shard on undesired node by allocating it, then setting `cluster.routing.allocation.enable=none` and  updating index setting to `routing.allocation.include=some-other-node-id`.

arteam replied:
I've decided to move away from the integration test to a unit tests where we can reliably randomly assign node ids and make consistent assertions. 

We already have the REST API test that kind of duplicates the previous integration test logic.

arteam replied:
I've decided to move away from the integration test to a unit tests where we can reliably randomly assign node ids and make consistent assertions. 

We already have the REST API test that kind of duplicates the previous integration test logic.

arteam committed:
node_is_desired should be true

arteam committed:
Update server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java

Co-authored-by: Ievgen Degtiarenko <ievgen.degtiarenko@gmail.com>

arteam committed:
Check cluster.routing.allocation.type is desired_balance

arteam committed:
Test error messages

arteam committed:
Make sure stats aren't negative

arteam committed:
Remove the integration test and move it to the unit test

arteam committed:
Revert "Make sure stats aren't negative"

This reverts commit 093251d583dcdac5e69784d734170c2ac7b566c7.

arteam committed:
Also handle unassigned shards

arteam commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,132 @@
+            Map<Integer, DesiredBalanceResponse.DesiredShards> indexDesiredShards = new HashMap<>();
+            for (int shardId = 0; shardId < indexRoutingTable.size(); shardId++) {
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(shardId);
+                // Do replicas overwrite primary shards here? Maybe we should we return
```
I kind of think that we do need to change the JSON output to return the primary along with replicas for every shard id.

arteam committed:
Return replica shards for desired balance

arteam committed:
Return the primary shard before replicas

arteam committed:
Don't depend on node_ids in REST API tests

We run them in different environments (single and multi-node)

arteam committed:
Add a comment about testing node ids

arteam committed:
Revert "Return the primary shard before replicas"

This reverts commit d864f59e20d28c59d73c2b181ab5cec406556483.

arteam committed:
Account for nullable allocationId and node

arteam committed:
Add an integration test

arteam committed:
Dynamically generate data to index

arteam committed:
Make the REST API test more reliable

Can't check the state and node desired status because we don't control
the ordering of shards

arteam committed:
Add tests for empty desired nodes

arteam committed:
Test for single shards

arteam commented:
@elasticmachine update branch

idegtiarenko commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_balance/10_basic.yml:
```diff
@@ -0,0 +1,136 @@
+            number_of_shards: 1
+            number_of_replicas: 0
+
+  - do:
+      bulk:
+        index: test
+        refresh: true
+        body:
+          - '{ "index": {"_id": "1"}}'
+          - '{  "field": "foo" }'
+          - '{ "index": {"_id": "2"}}'
+          - '{ "field": "foo 2" }'
+          - '{ "index": {"_id": "3"}}'
+          - '{ "field": "foo 3" }'
+          - '{ "index": {"_id": "4"}}'
+          - '{ "field": "foo 4" }'
+          - '{ "index": {"_id": "5"}}'
+          - '{ "field": "bar" }'
```
NIT: this should not affect desired balance

DaveCTurner replied:
:+1: I don't think we need to index any docs in these tests, and would prefer not to do so.

arteam replied:
Done!

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/DesiredBalanceResponse.java:
```diff
@@ -0,0 +1,212 @@
+    public static DesiredBalanceResponse from(StreamInput in) throws IOException {
+        return new DesiredBalanceResponse(
+            DesiredBalanceStats.readFrom(in),
+            in.readMap(StreamInput::readString, v -> v.readMap(in1 -> in1.readVInt(), DesiredShards::from))
```
NIT: Could we use `readImmutableMap` for both top level and nested map here?

idegtiarenko commented on server/src/test/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceActionIntegrationTests.java:
```diff
@@ -0,0 +1,154 @@
+
+        indexData(index);
+
+        // Wait until all replicas have started
+        waitUntil(() -> {
+            IndexRoutingTable indexRoutingTable = client().admin().cluster().prepareState().get().getState().routingTable().index(index);
+            for (int i = 0; i < indexRoutingTable.size(); i++) {
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(i);
+                for (int j = 0; j < shardRoutingTable.size(); j++) {
+                    ShardRouting shard = shardRoutingTable.shard(j);
+                    if (shard.state() != ShardRoutingState.STARTED) {
+                        return false;
+                    }
+                }
+            }
+            return true;
+        });
```
This should be enforced by a `createIndex`.
Was it not the case for some of the test runs?

idegtiarenko replied:
Uh, my bad, we only wait for primaries. Create with `waitForActiveShards=ALL` will do this

DaveCTurner replied:
Or else use the health API to wait for a green status.

arteam replied:
No, but I was able to achieve this was David's advice of using the cluster health API :+1:

DaveCTurner commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_balance/10_basic.yml:
```diff
@@ -0,0 +1,136 @@
+  - gte: { stats.computation_time_in_millis: 0 }
+  - gte: { stats.reconciliation_time_in_millis: 0 }
+
+  - match: { routing_table.test.0.current.0.state: 'STARTED' }
+  - match: { routing_table.test.0.current.0.shard_id: 0 }
+  - match: { routing_table.test.0.current.0.index: test }
+  - is_true: 'routing_table.test.0.current.0.node_is_desired'
+  - is_false: 'routing_table.test.0.current.0.relocating_node'
+  - is_false: 'routing_table.test.0.current.0.relocating_node_is_desired'
+  - match: { routing_table.test.0.desired.total: 1 }
+  - gte: { routing_table.test.0.desired.unassigned: 0 }
+  - gte: { routing_table.test.0.desired.ignored: 0 }
+  - is_true: 'routing_table.test.0.desired.node_ids'
```
Might these assertions be flaky? Should we first wait for the cluster to be `green` with no initializing/relocating shards first?

idegtiarenko replied:
I am not sure we have access to started node ids in yaml tests, do we?

arteam replied:
Done!

DaveCTurner commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_balance/10_basic.yml:
```diff
@@ -0,0 +1,136 @@
+  # whether the test is run on a single or multi-node cluster. Also can't match the status and state
+  # because we can't control the ordering of replicas and primaries in the output
+
+  - match: { routing_table.test.0.current.0.state: /STARTED|INITIALIZING|UNASSIGNED/ }
+  - match: { routing_table.test.0.current.0.shard_id: 0 }
+  - match: { routing_table.test.0.current.0.index: test }
+  - is_false: 'routing_table.test.0.current.0.relocating_node'
+  - is_false: 'routing_table.test.0.current.0.relocating_node_is_desired'
+  - match: { routing_table.test.0.current.1.state: /STARTED|INITIALIZING|UNASSIGNED/ }
+  - match: { routing_table.test.0.current.1.shard_id: 0 }
+  - match: { routing_table.test.0.current.1.index: test }
+  - is_false: 'routing_table.test.0.current.1.relocating_node'
+  - is_false: 'routing_table.test.0.current.1.relocating_node_is_desired'
+  - match: { routing_table.test.0.desired.total: 2 }
+  - gte: { routing_table.test.0.desired.unassigned: 0 }
+  - gte: { routing_table.test.0.desired.ignored: 0 }
+  - is_true: 'routing_table.test.0.desired.node_ids'
+
+  - match: { routing_table.test.1.current.0.state: /STARTED|INITIALIZING|UNASSIGNED/ }
+  - match: { routing_table.test.1.current.0.shard_id: 1 }
+  - match: { routing_table.test.1.current.0.index: test }
+  - is_false: 'routing_table.test.1.current.0.relocating_node'
+  - is_false: 'routing_table.test.1.current.0.relocating_node_is_desired'
+  - match: { routing_table.test.1.current.1.state: /STARTED|INITIALIZING|UNASSIGNED/ }
+  - match: { routing_table.test.1.current.1.shard_id: 1 }
+  - match: { routing_table.test.1.current.1.index: test }
+  - is_false: 'routing_table.test.1.current.1.relocating_node'
+  - is_false: 'routing_table.test.1.current.1.relocating_node_is_desired'
+  - match: { routing_table.test.1.desired.total: 2 }
+  - gte: { routing_table.test.1.desired.unassigned: 0 }
+  - gte: { routing_table.test.1.desired.ignored: 0 }
+  - is_true: 'routing_table.test.1.desired.node_ids'
```
Likewise here - might these be flaky? Should we first wait for the cluster to be `green` with no initializing/relocating shards first?

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/DesiredBalanceRequest.java:
```diff
@@ -0,0 +1,27 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+package org.elasticsearch.action.admin.cluster.desiredbalance;
```
The rest of this code is just in the allocation package, we're not (currently) giving it its own package, so I suggest doing the same here (applies to all the classes in this new package)

```suggestion
package org.elasticsearch.action.admin.cluster.allocation;
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -705,6 +708,8 @@ public <Request extends ActionRequest, Response extends ActionResponse> void reg
         actions.register(UpdateHealthInfoCacheAction.INSTANCE, UpdateHealthInfoCacheAction.TransportAction.class);
         actions.register(FetchHealthInfoCacheAction.INSTANCE, FetchHealthInfoCacheAction.TransportAction.class);
 
+        actions.register(GetDesiredBalanceAction.INSTANCE, TransportGetDesiredBalanceAction.class);
```
Likewise here I suggest putting this alongside the allocation explain action.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/ActionModule.java:
```diff
@@ -887,6 +892,8 @@ public void initRestHandlers(Supplier<DiscoveryNodes> nodesInCluster) {
         registerHandler.accept(new RestUpdateDesiredNodesAction());
         registerHandler.accept(new RestDeleteDesiredNodesAction());
 
+        registerHandler.accept(new RestGetDesiredBalanceAction());
```
Likewise here I suggest putting this alongside the allocation explain action.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -399,6 +394,7 @@ protected void configure() {
         bind(TaskResultsService.class).asEagerSingleton();
         bind(AllocationDeciders.class).toInstance(allocationDeciders);
         bind(ShardsAllocator.class).toInstance(shardsAllocator);
+        bind(DesiredBalanceShardsAllocator.class).toInstance(desiredBalanceShardsAllocator);
```
I don't think we want to create this if it's not configured. Instead we should use the instance provided for `ShardsAllocator.class`, and try to cast it to a `DesiredBalanceShardsAllocator` where necessary.

DaveCTurner reviewed (commented):
Looks good, I left a few minor comments.

arteam committed:
Remove indexing data (which is unnecessary for desired balance)

arteam committed:
Run multi-shard cluster only on multiple nodes

arteam committed:
Read immutable maps from serialization

arteam committed:
Check for the status of shards via the cluster health API

arteam committed:
Move to the allocation package

arteam committed:
Run spotless

arteam committed:
Run spotless for tests

elasticsearchmachine commented:
Hi @arteam, I've created a changelog YAML for you.

arteam committed:
Update docs/changelog/91038.yaml

arteam commented:
@elasticmachine update branch

arteam committed:
We don't need to expose DesiredBalanceAllocator via Guice

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,138 @@
+        ClusterState state,
+        ActionListener<DesiredBalanceResponse> listener
+    ) throws Exception {
+        String allocatorName = ClusterModule.SHARDS_ALLOCATOR_TYPE_SETTING.get(state.metadata().settings());
+        if (allocatorName.equals(ClusterModule.DESIRED_BALANCE_ALLOCATOR) == false) {
+            listener.onFailure(
+                new ResourceNotFoundException(
+                    "Expected the shard balance allocator to be `desired_balance`, but got `" + allocatorName + "`"
+                )
+            );
+            return;
+        }
+
+        DesiredBalanceShardsAllocator desiredBalanceShardsAllocator = (DesiredBalanceShardsAllocator) shardsAllocator;
```
nit/followup: I'd prefer that the `shardsAllocator` field was of type `DesiredBalanceShardsAllocator` and `@Nullable`, and that we do this cast (conditionally) in the constructor. That way there's no need to look at the allocator type setting here, you can just test whether the field is `null` or not, and there's no need to worry about future changes that might break the assumed relationship between the setting and the actual allocator type.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,138 @@
+            DesiredBalanceRequest::new,
+            indexNameExpressionResolver,
+            DesiredBalanceResponse::from,
+            ThreadPool.Names.SAME
```
Nit/followup: iterating over the whole routing table could be somewhat nontrivial, let's fork this elsewhere just to be safe:

```suggestion
            ThreadPool.Names.MANAGEMENT
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestGetDesiredBalanceAction.java:
```diff
@@ -0,0 +1,41 @@
+        return restChannel -> client.execute(
+            GetDesiredBalanceAction.INSTANCE,
+            new DesiredBalanceRequest(),
+            new RestToXContentListener<>(restChannel)
```
Nit/followup: the response could be pretty sizeable here, I think we should use `RestChunkedToXContentListener` here so that we respond in a streaming fashion instead.

DaveCTurner reviewed (approved):
LGTM, I left some nits but we can treat them as (very minor) bugs to be addressed in follow-ups.

henningandersen reviewed (commented):
Looks good, no need to await my review before merging.

arteam committed:
Update server/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportGetDesiredBalanceAction.java

Co-authored-by: David Turner <david.turner@elastic.co>

arteam committed:
Move null checks for desiredBalanceShardsAllocator to constructor

arteam committed:
Update API tip for desired Balance
